src/webgpu/api/operation/sampling/anisotropy.spec.ts

```diff
@@ -6,14 +6,377 @@ Note that anisotropic filtering is never guaranteed to occur, but we might be ab
+  .desc(
+    `anisotropic filter rendering tests that draws a slanted plane and samples from a texture
+    containing mipmaps of different colors.
+    A similar webgl demo is at https://jsfiddle.net/t8k7c95o/5/`
+  )
+  .params([
+    {
+      maxAnisotropy: 1,
+      _results: [
+        {
+          coord: { x: xMiddle, y: 2 },
+          expected: colors[2],
+        },
+        {
+          coord: { x: xMiddle, y: 6 },
+          expected: [colors[0], colors[1]],
+        },
+      ],
+    },
+    {
+      maxAnisotropy: 2,
+      _results: [
+        {
+          coord: { x: xMiddle, y: 2 },
+          expected: [colors[1], colors[2]],
+        },
+        {
+          coord: { x: xMiddle, y: 6 },
+          expected: colors[0],
+        },
+      ],
+    },
+    {
+      maxAnisotropy: 16,
+      _results: [
+        {
+          coord: { x: xMiddle, y: 2 },
+          expected: [colors[0], colors[1]],
+        },
+        {
+          coord: { x: xMiddle, y: 6 },
+          expected: colors[0],
+        },
+      ],
+    },
+  ])
+  .fn(async t => {
+    // init texture
+    const mipLevelCount = 3;
+    const textureSizeMipmap0 = 1 << (mipLevelCount - 1);
+    const texture = t.device.createTexture({
+      mipLevelCount,
+      size: { width: textureSizeMipmap0, height: textureSizeMipmap0, depth: 1 },
+      format: kTextureFormat,
+      usage: GPUTextureUsage.COPY_DST | GPUTextureUsage.SAMPLED,
+    });
+
+    const textureEncoder = t.device.createCommandEncoder();
+    // Populate each level with a different color
+    for (let i = 0; i < mipLevelCount; i++) {
+      const mipmapSize = mipSize([textureSizeMipmap0], i)[0];
+      const bufferSize = kBytesPerRow * mipmapSize; // RGBA8 for each pixel (256 > 16 * 4)
+
+      // init texture data
+      const data: Uint8Array = new Uint8Array(bufferSize);
+      const color = colors[i];
+      for (let r = 0; r < mipmapSize; r++) {
+        const o = r * kBytesPerRow;
+        for (let c = o, end = o + mipmapSize * 4; c < end; c += 4) {
+          data[c] = color[0];
+          data[c + 1] = color[1];
+          data[c + 2] = color[2];
+          data[c + 3] = color[3];
+        }
+      }
+      const buffer = t.makeBufferWithContents(
+        data,
+        GPUBufferUsage.COPY_SRC | GPUBufferUsage.COPY_DST
+      );
+      const bytesPerRow = kBytesPerRow;
+      const rowsPerImage = mipmapSize;
+
+      textureEncoder.copyBufferToTexture(
+        {
+          buffer,
+          bytesPerRow,
+          rowsPerImage,
+        },
+        {
+          texture,
+          mipLevel: i,
+          origin: [0, 0, 0],
+        },
+        [mipmapSize, mipmapSize, 1]
+      );
+    }
+    t.device.defaultQueue.submit([textureEncoder.finish()]);
+
+    const textureView = texture.createView();
+
+    const sampler = t.device.createSampler({
+      magFilter: 'linear',
+      minFilter: 'linear',
+      mipmapFilter: 'linear',
+      maxAnisotropy: t.params.maxAnisotropy,
```

Comment:
Commenting out this line should not cause test failures, as it's valid for implementations to ignore this parameter. I guess the best way to achieve this is to change errors to warnings? Unfortunately this is not currently doable with the expect functions :(

Maybe it's possible to loosen the expectations? (maybe have both warnings and hard errors)

Reply:
What about expecting the result to be equal to either that of `maxAnisotropy = t.params.maxAnisotropy` or `maxAnisotropy = 1`? (which is assuming if device support anisotropy, it supports up to 16)

Reply:
> (which is assuming if device support anisotropy, it supports up to 16)

IIRC there are a few devices out there that only support anisotropy up to 4, unfortunately. Though I suppose we could check 1 and 4 instead of 16.

> What about expecting the result to be equal to either that of `maxAnisotropy = t.params.maxAnisotropy` or `maxAnisotropy = 1`?

I don't quite understand this, expecting the result of what to be equal to one of those? Isn't the "result" the case with maxAnisotropy=t.params.maxAnisotropy?

Unfortunately since there are no guarantees about the implementation of anisotropic filtering (and the fact that the Vulkan spec's reference implementation is really complicated), probably the best we can do here is warnings. Warnings would be strong enough to guarantee that WebGPU implementations are correct in automated testing - and it's not within-spec to strictly test hardware behavior.

If the tests can only produce warnings anyway, then it's okay to warn on hardware that implements anisotropic filtering poorly (e.g. <16, or in a substandard way) or not at all.

Reply:
Now testing maxAnisotropy = [1, 4]. And make the maxAnisotorpy=4 case to generate warning when fails
"Unimportant: I thought I mean when maxAnisotropy = 4, the result could be equal to that on 1 or 4 on a device with full support of anisotoropic rendering"